### PR TITLE
Ability to Highlight Multiple Search Terms #5220

### PR DESCRIPTION
**Pull request by michaelhsnyder**
If PDF document was part of a document search would like the ability to highlight multiple simultaneous terms and/or phrases that were returned as a search result. Currently there is no way to do this.

Change allows search terms to be passed in via URL parameter (search = terms list). The terms will all be highlighted with the same classes used in the findbar. All instances of all terms will be highlighted. Each term must be separated by a space (%20) and each phrase must be separated with an equals (%3d). The document will scroll to the match nearest to the top of the page after loading. Would like the scroll to be animated, but too much effort without another library or jquery at this time.

Examples:

unix dos windows =>  &search=unix%20dos%20windows
"sales manager" McDonalds => &search=sales%3dmanager%20McDonalds 

Also allows terms to be sent in via parent URL if using iframes and/or multiple pdf viewers on  same search result. This allows the iframes to load different documents without having to append the search terms each time.

### CODE REVIEW

File: web/text_layer_builder.js [hunk 1]
```diff
@@ -363,19 +369,21 @@ var TextLayerBuilder = (function TextLayerBuilderClosure() {
...
-        var match = matches[i];
-        var begin = Math.max(clearedUntilDivIdx, match.begin.divIdx);
-        for (var n = begin, end = match.end.divIdx; n <= end; n++) {
-          var div = textDivs[n];
-          div.textContent = bidiTexts[n].str;
-          div.className = '';
-        }
-        clearedUntilDivIdx = match.end.divIdx + 1;
-      }
+      if (PDFJS.multiple === undefined) {
```
Comment (speedplane) [inline web/text_layer_builder.js]: Why are we skipping this code if `multiple` is defined?

File: web/pdf_find_controller.js [hunk 2]
```diff
@@ -377,3 +407,60 @@ var PDFFindController = (function PDFFindControllerClosure() {
...
+            //also check parent window for any search terms
+            if (params[paramname] === undefined) {
+                loc = parent.document.location.href.replace('#', '&');
+                loc = loc.replace(/\+/ig, '%20');
+                params = PDFView.parseQueryString(loc);
+            }
+
+            if (params[paramname] !== undefined) {
+                var parambase = params[paramname].replace(/\"/ig, '');
+                PDFJS.multiple = parambase.split(/\s+/);
```
Comment (speedplane) [inline web/pdf_find_controller.js]: Maybe consider a better name than `multiple`.... multiple what? Maybe a better one would be `multipleSearchParam`?

File: web/pdf_find_controller.js [hunk 3]
```diff
@@ -228,6 +245,7 @@ var PDFFindController = (function PDFFindControllerClosure() {
         this.resumePageIdx = null;
         this.pageMatches = [];
         var self = this;
+        var qwy = '';
```
Comment (speedplane) [inline web/pdf_find_controller.js]: What's the purpose of this local var? Why not just set `self.state.query` directly?

File: web/pdf_find_controller.js [hunk 4]
```diff
@@ -203,15 +203,32 @@ var PDFFindController = (function PDFFindControllerClosure() {
...
         page.textLayer.updateMatches();
       }
+
+      //scroll match into view
+      if (PDFJS.multiple !== undefined) {
+          try {
+              var d = document.getElementsByClassName('highlight');
+              if (d.length > 0) {
+                  var vc = document.getElementById('viewerContainer');
+                  if (vc.scrollTop < 40) {
```
Comment (speedplane) [inline web/pdf_find_controller.js]: Is the `40` here supposed to be toolbar height? That should probably not be set as a constant here.

**Comment by willspecht** [comment]
Just a note, this implementation is broken.  If you have multiple search terms that appear in the same div only the last search term is highlighted.  I fixed this by generating all matches from the query string then updating the highlighting once, instead of updating once per search term.  Should I commit these changes to this branch or create a separate branch?

**Comment by timvandermeij** [comment]
Since this PR has not been updated for a long time, it would be most helpful if you could create a new PR with the changes from this PR and your own fixes combined. Be sure to squash everything into one commit (see https://github.com/mozilla/pdf.js/wiki/Squashing-Commits) and set the original author as the author of the commit using `--author="mainseq <ron@mainsequence.net>"` when committing. Thank you!

**Comment by Snuffleupagus** [comment]
@timvandermeij Wasn't this PR superseded by #5579?

**Comment by timvandermeij** [comment]
You are right. I'm closing this as superseded by #5579. Refer to https://github.com/mozilla/pdf.js/pull/5579#issuecomment-68028444 and the comment below it for the disadvantages of this approach and the advantages of the approach in that PR. That PR is also more maintained.

**Pull request closed without merging**